From b6aef99eefb947efe093d3d5742da98661d41b35 Mon Sep 17 00:00:00 2001
From: Yurii Rashkovskii <me@yrashk.com>
Date: Mon, 29 Oct 2018 20:59:34 -0700
Subject: [PATCH] Problem: noi-wiki relies on a mutation observer

noi-wiki component wants to know when it's details were changed and it's
using a mutation observer for that. However, this is a bit heavy-handed.

Solution: create a singular document mutation observer
that emits its own even specifically when a wiki is updated.
---
 src/core/events.noi   | 26 ++++++++++++++++++++++++++
 src/core/noi-wiki.noi | 16 +++++-----------
 2 files changed, 31 insertions(+), 11 deletions(-)
 create mode 100644 src/core/events.noi

diff --git a/src/core/events.noi b/src/core/events.noi
new file mode 100644
index 0000000..adedaf4
--- /dev/null
+++ b/src/core/events.noi
@@ -0,0 +1,26 @@
+<noi-wiki id="noi/core/events" content-type="text/x.html-source">
+  <script>
+  document.addEventListener('noi-render-ready', () => {
+  new MutationObserver((mutations) => {
+      var updates = {};
+      for (var mutation of mutations) {
+        if (mutation.target.tagName.toLowerCase() === "noi-wiki") {
+           if (mutation.type == "childList")  {
+               updates[mutation.target.id] = Object.assign(updates[mutation.target.id] || {}, {
+                 wiki: mutation.target,
+               });
+           }
+           if (mutation.type == "attributes" && mutation.attributeName !== "class" && !mutation.attributeName.startsWith("data-")) {
+               updates[mutation.target.id] = Object.assign(updates[mutation.target.id] || {}, {
+                 wiki: mutation.target,
+               });
+           }
+        }
+      }
+      for (var update in updates) {
+          document.dispatchEvent(new CustomEvent('noi-wiki-updated', {detail: updates[update]}));
+      }
+    }).observe(document.body, {childList: true, attributes: true, subtree: true});
+  });
+  </script>
+</noi-wiki>
diff --git a/src/core/noi-wiki.noi b/src/core/noi-wiki.noi
index 2e806b8..43baa90 100644
--- a/src/core/noi-wiki.noi
+++ b/src/core/noi-wiki.noi
@@ -3,23 +3,17 @@
     class NoiWiki extends HTMLElement {
       constructor() {
         super();
-        this.attachShadow({ mode: 'open' });
-        this.observer = new MutationObserver((e) => {
-          this.render();
+        document.addEventListener('noi-wiki-updated', (e) => {
+          if (e.detail.wiki == this) {
+            this.render();
+          }
         });
+        this.attachShadow({ mode: 'open' });
       }
       connectedCallback() {
         this.shouldRender = this.hasAttribute("data-rendered");
-        this.observer.observe(this, {
-          attributes: true,
-          childList: true,
-          subtree: false
-        });
         this.render();
       }
-      disconnectedCallback() {
-        this.observer.disconnect();
-      }
       static get observedAttributes() {
         return ['data-rendered'];
       }
-- 
2.18.1

